win32ss/user/user32/windows/cursoricon.c

```diff
@@ -1148,6 +1149,16 @@ BITMAP_LoadImageW(
         goto end;
     CopyMemory(pbmiCopy, pbmi, iBMISize);
 
+    /* Test if MSVC Resource Compiler used and if so, adjust pvBits */
```

Comment:
My question is this, actually: Do you really _**understand**_ why the MSVC resource compiler does that differently?
Isn't it, actually, that it aligned the data, or something?

Reply:
This could be alignment indeed.

Reply:
Aren't these 12 bytes simply the bitfields, i.e. they would be present in any BI_BITFIELDS image?
It seems to essentially be a "palette" that tells you which bits in each pixel specify R vs G vs B, respectively: https://www.virtualdub.org/blog2/entry_177.html

Reply:
> Aren't these 12 bytes simply the bitfield mask

Correct, it is the mask. It has nothing to do with MSVC RC, it is simply a bitmap with the unofficial V2 header created by Photoshop.

The real bug is in the `max` macro in `bitmap_info_size`. Allow me to quote from MSDN about biSize:

> Specifies the number of bytes required by the structure. This value does not include the size of the color table **or the size of the color masks, if they are appended to the end of structure**.

Reply:
No. I do NOT understand why the MSVC compiler does this. If it is an alignment issue it seems odd because we are dealing with 32 bits per plane which means that each pixel's data is 32 bits and this makes it seem that 12 additional bytes for alignment would not be correct. But maybe this considers an offset for the header or other reason, so that the in-memory file mapped data is actually aligned. The 12 byte bitfield masks are included within the header as shown by the example below. The extra 12 bytes that I am accounting for are beyond the end of the header in our case.

![BitFields](https://github.com/reactos/reactos/assets/32620577/7a8c28c6-2541-4934-b61a-5d2263f099d4)

As far as someone must know about how to handle these, Resource Hacker's latest version of 5.2.4 is not able to display these embedded images from either an MSVC compiled or GCC compiled DLL in its 'Editor View' pane where bitmaps are normally shown. We have example bitmaps of V4 in comctl32.dll (images 1-6) and V5 in browseiw.dll (index 240) and the 12 bytes size difference is the same between these. V4 has a header size of 0x6c and V5, of 0x7C, so this seems to point to this difference NOT being alignment related. Below you can see that the size in the bottom left corner for GCC is shown as 0xF27C and for MSVC it is 0xF288 so there is the 12 byte difference.

![browseui-gcc-msvc-compare](https://github.com/reactos/reactos/assets/32620577/525d4ff9-30d8-426c-a59d-c729fe2272ef)

Thanks for everyone's comments and interest here.

Reply:
An interesting observation is that the comctl32.dll from Win7 can be viewed in Resource Hacker just fine. All bitmaps can be selected and viewed in the 'Editor View'. All of the images shows 'bpp: 4; hdr:v1', so these are much less difficult to handle. This makes it seem that ReactOS may be overly complicating this implementation.

![Win7_ResHacker_comctl32_dll](https://github.com/reactos/reactos/assets/32620577/bba57c50-1344-42e9-8f7f-8790a40b2c5b)

Reply:
I wonder also whether the toolbar bitmap in question that we have, is actually valid...

Reply:
All indications that I have seen show that they are valid. You can open them with Paint and Windows Photo Viewer. Also, just in explorer you can select each *.bmp file and the information pane at the bottom will show you a small picture and the width and height information.

Reply:
> > Aren't these 12 bytes simply the bitfield mask
> 
> Correct, it is the mask. It has nothing to do with MSVC RC.
> 
> The real bug is in the `max` macro in `bitmap_info_size`. Allow me to quote from MSDN about biSize:
> 
> > Specifies the number of bytes required by the structure. This value does not include the size of the color table **or the size of the color masks, if they are appended to the end of structure**.
> 
> Note: The unofficial(?) V2 and V3 bitmap header formats and the official V4 and V5 do include the masks in biSize. If `biSize > FIELD_OFFSET(BITMAPV4HEADER, bV4BlueMask)` then the masks are included, otherwise you must add them.

That makes sense

Reply:
The bug in `bitmap_info_size` I already mentioned seems to be a separate issue.

It does seem like there is a bug in how MSVC RC handles `BITMAPV4/5HEADER`+`BI_BITFIELDS`.

 - As indicated by Doug-Lyons, the bitmap resource is 12 bytes too large and those 12 bytes appear after the bitmap header.
 - If I change `biSize` of browseui's 240.bmp to `sizeof(BITMAPINFOHEADER)`, RC handles it correctly and `LoadImage` will work.
 - If I change the compression to BI_RG, RC handles it correctly but `LoadImage` will fail.
 
----
 - I used a pretty old MSVC RC version, someone should test on the latest version on Visual Studio.
 - I believe `LoadImage` on Windows only supports `BITMAPCOREHEADER` and `BITMAPINFOHEADER` bitmaps (Gimp [had issues](https://gitlab.gnome.org/GNOME/gimp/-/issues/2010) with this in the past).

Reply:
Another way of interpreting this:

Before I get to the quote, keep in mind that BITMAPINFO is
```C
typedef struct tagBITMAPINFO {
  BITMAPINFOHEADER bmiHeader;
  RGBQUAD          bmiColors[ANYSIZE_ARRAY];
} BITMAPINFO, *LPBITMAPINFO, *PBITMAPINFO;
```

The documentation for BITMAPV5HEADER says

> If the bV5Compression member of the BITMAPV5HEADER is BI_BITFIELDS, the bmiColors member contains three DWORD color masks that specify the red, green, and blue components of each pixel.

And the [overview of headers](https://learn.microsoft.com/en-us/windows/win32/gdi/bitmap-header-types):

 > ...The red, green, and blue bitfield masks for BI_BITFIELD bitmaps immediately follow the BITMAPINFOHEADER, BITMAPV4HEADER, and BITMAPV5HEADER structures. 

I believe they are talking about packed DIBs here but perhaps they are storing the resource like this as well? Or everyone is storing BITMAPV4/5HEADERs incorrectly in .bmp files by not storing the mask twice?

Reply:
After looking into this in more detail, I now agree that the bitmap used in the comctl32.dll for the Wordpad 'Open File' dialog toolbar is named 'idb_view_small.bmp' and these 12 bytes that are contained after the BITMAPV4HEADER (size 0x6C) are the bitfield masks, present in any BI_BITFIELDS image that have been duplicated when compiling with 'rc.exe', the Resource Compiler that is part of the MSVC compiler. These seem to be included, but do not seem to be significant in any way. To prove this, I took the 'idb_view_small.bmp' image and edited the normal 12 bytes of bitfield masks replacing the zeroes with four different values (0x00, 0x55, 0xAA, and 0xFF). Here is an example in Frhed showing the bytes as 0x55.

![idb_view_small_55](https://github.com/reactos/reactos/assets/32620577/d9b675d7-93d3-46ab-9149-b8ce19cd564e)

Then I compiled these with both MSVC and GCC and swapped out the comctl32.dll in each case with the one in a working bootcd install and opened Wordpad and then opened the "Open File" dialog box. There was no difference in the look of this dialog box with any of these changes. I proved that the bytes are actually duplicated by opening each newly compiled comctl32.dll and comparing the difference between the MSVC and GCC versions. Here are their results:

![comctl32_compare_10](https://github.com/reactos/reactos/assets/32620577/3f576791-715f-4dfb-bf56-ac6a3ade9f4c)

![comctl32_compare_12](https://github.com/reactos/reactos/assets/32620577/90a37c60-e92b-4ec0-adb5-c0e74f53fda8)

![comctl32_compare_11](https://github.com/reactos/reactos/assets/32620577/fdbf304c-78fe-4ee5-951f-0aaccde2545a)

![comctl32_compare_09](https://github.com/reactos/reactos/assets/32620577/976c1b0a-ad54-42db-9843-fa8ac8adc675)

So this seems to show that though these bytes get duplicated, that they are not really significant for us.

Reply:
Upon further investigation, I have proven to myself that this same adjustment is needed for 16-bpp BI_BITFIELDS bitmaps. GCC's windres.exe handles this type of resource correctly adding the extra 12 bytes as does MSVC's rc.exe. More details to follow soon. Thanks.